### PR TITLE
Use CUnicode instead of Unicode

### PR DESCRIPTION
**Pull request by martinRenou**
Since Traitlets 5, setting a Unicode trait from command line to be empty string "" is not supported anymore. It is only supported when using the CUnicode class.

Setting a command line argument to an empty list is needed in Voila.

Related to https://github.com/voila-dashboards/voila/pull/699

### CODE REVIEW

**Comment by martinRenou** [comment]
cc. @maartenbreddels

**Comment by blink1073** [comment]
The test failures are related to https://github.com/jupyter/nbconvert/issues/1366

File: jupyter_server/serverapp.py [hunk 1]
```diff
@@ -782,7 +778,7 @@ def _token_default(self):
     def _token_changed(self, change):
         self._token_generated = False
 
-    password = Unicode(u'', config=True,
+    password = CUnicode(u'', config=True,
```
Comment (jasongrout) [inline jupyter_server/serverapp.py:781]: With python 3, we don't need the u prefix here and elsewhere
```suggestion
    password = CUnicode('', config=True,
```

File: jupyter_server/serverapp.py [hunk 2]
```diff
@@ -662,7 +658,7 @@ def _valdate_ip(self, proposal):
             value = u''
         return value
 
-    custom_display_url = Unicode(u'', config=True,
+    custom_display_url = CUnicode(u'', config=True,
```
Comment (jasongrout) [inline jupyter_server/serverapp.py:661]:
```suggestion
    custom_display_url = CUnicode('', config=True,
```

File: jupyter_server/serverapp.py [hunk 3]
```diff
@@ -683,19 +679,19 @@ def _valdate_ip(self, proposal):
         help=_("The number of additional ports to try if the specified port is not available.")
     )
 
-    certfile = Unicode(u'', config=True,
+    certfile = CUnicode(u'', config=True,
```
Comment (jasongrout) [inline jupyter_server/serverapp.py:682]:
```suggestion
    certfile = CUnicode('', config=True,
```

File: jupyter_server/serverapp.py [hunk 4]
```diff
@@ -683,19 +679,19 @@ def _valdate_ip(self, proposal):
...
         help=_("""The full path to an SSL/TLS certificate file.""")
     )
 
-    keyfile = Unicode(u'', config=True,
+    keyfile = CUnicode(u'', config=True,
```
Comment (jasongrout) [inline jupyter_server/serverapp.py:686]:
```suggestion
    keyfile = CUnicode('', config=True,
```

File: jupyter_server/serverapp.py [hunk 5]
```diff
@@ -683,19 +679,19 @@ def _valdate_ip(self, proposal):
...
         help=_("""The full path to a private key file for usage with SSL/TLS.""")
     )
 
-    client_ca = Unicode(u'', config=True,
+    client_ca = CUnicode(u'', config=True,
```
Comment (jasongrout) [inline jupyter_server/serverapp.py:690]:
```suggestion
    client_ca = CUnicode('', config=True,
```

File: jupyter_server/serverapp.py [hunk 6]
```diff
@@ -897,7 +893,7 @@ def _default_allow_remote(self):
                         (ServerApp.browser) configuration option.
                         """)
 
-    browser = Unicode(u'', config=True,
+    browser = CUnicode(u'', config=True,
```
Comment (jasongrout) [inline jupyter_server/serverapp.py:896]:
```suggestion
    browser = CUnicode('', config=True,
```

**Comment by martinRenou** [comment]
I've only updated this file, maybe it would make sense to make the same change in other places?

**Comment by kevin-bates** [comment]
Hi @martinRenou,

Looking at this via `--ServerApp.token=""`, I'm not finding an issue with traitlets 5 (`5..04`).  Here's output of two invocations; the first with traitlets `4.3.3.`, the second with traitlets `5.0.4` - both using `master`.  Could you please clarify how the issue with traitlets 5 manifests itself?

```
jupyter_server (master)$ jupyter-server --ServerApp.token=""
[I 2020-09-17 17:27:17.363 ServerApp] elyra | extension was found and enabled by nbclassic. Consider moving the extension to Jupyter Server's extension paths.
[I 2020-09-17 17:27:17.363 ServerApp] elyra | extension was successfully linked.
[I 2020-09-17 17:27:17.366 ServerApp] jupyterlab_git | extension was found and enabled by nbclassic. Consider moving the extension to Jupyter Server's extension paths.
[I 2020-09-17 17:27:17.366 ServerApp] jupyterlab_git | extension was successfully linked.
[I 2020-09-17 17:27:17.382 ServerApp] nbdime | extension was found and enabled by nbclassic. Consider moving the extension to Jupyter Server's extension paths.
[I 2020-09-17 17:27:17.382 ServerApp] nbdime | extension was successfully linked.
[I 2020-09-17 17:27:17.382 ServerApp] nbclassic | extension was successfully linked.
[W 2020-09-17 17:27:17.424 ServerApp] All authentication is disabled.  Anyone who can connect to this server will be able to run code.
[I 2020-09-17 17:27:17.425 ServerApp] The port 8888 is already in use, trying another port.
[I 2020-09-17 17:27:17.432 ServerApp] elyra | extension was successfully loaded.
[I 2020-09-17 17:27:17.433 ServerApp] jupyterlab_git | extension was successfully loaded.
[I 2020-09-17 17:27:17.438 ServerApp] nbclassic | extension was successfully loaded.
[I 2020-09-17 17:27:17.535 ServerApp] nbdime | extension was successfully loaded.
[I 2020-09-17 17:27:17.535 ServerApp] Serving notebooks from local directory: /Users/kbates/repos/oss/gateway-experiments/jupyter_server
[I 2020-09-17 17:27:17.536 ServerApp] Jupyter Server 1.0.0rc16 is running at:
[I 2020-09-17 17:27:17.536 ServerApp] http://localhost:8889/
[I 2020-09-17 17:27:17.536 ServerApp]  or http://127.0.0.1:8889/
[I 2020-09-17 17:27:17.536 ServerApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
^C[I 2020-09-17 17:27:28.893 ServerApp] interrupted
Serving notebooks from local directory: /Users/kbates/repos/oss/gateway-experiments/jupyter_server
0 active kernels
Jupyter Server 1.0.0rc16 is running at:
http://localhost:8889/
 or http://127.0.0.1:8889/
Shutdown this Jupyter server (y/[n])? ^C[C 2020-09-17 17:27:29.341 ServerApp] received signal 2, stopping
[I 2020-09-17 17:27:29.342 ServerApp] Shutting down 0 kernels

jupyter_server (master)$ pip install --upgrade traitlets
Looking in indexes: https://pypi.org/simple/
Collecting traitlets
  Downloading https://files.pythonhosted.org/packages/e6/20/f73a1130598ffaf561fda74d761bbe66db7a3639f34456761e781509881b/traitlets-5.0.4-py3-none-any.whl (98kB)
     |████████████████████████████████| 102kB 1.9MB/s 
Requirement already satisfied, skipping upgrade: ipython-genutils in /opt/anaconda3/envs/jupyter-server-dev/lib/python3.7/site-packages (from traitlets) (0.2.0)
Installing collected packages: traitlets
  Found existing installation: traitlets 4.3.3
    Uninstalling traitlets-4.3.3:
      Successfully uninstalled traitlets-4.3.3
Successfully installed traitlets-5.0.4


jupyter_server (master)$ jupyter-server --ServerApp.token=""
[I 2020-09-17 17:27:45.167 ServerApp] elyra | extension was found and enabled by nbclassic. Consider moving the extension to Jupyter Server's extension paths.
[I 2020-09-17 17:27:45.167 ServerApp] elyra | extension was successfully linked.
[I 2020-09-17 17:27:45.169 ServerApp] jupyterlab_git | extension was found and enabled by nbclassic. Consider moving the extension to Jupyter Server's extension paths.
[I 2020-09-17 17:27:45.169 ServerApp] jupyterlab_git | extension was successfully linked.
[I 2020-09-17 17:27:45.179 ServerApp] nbdime | extension was found and enabled by nbclassic. Consider moving the extension to Jupyter Server's extension paths.
[I 2020-09-17 17:27:45.179 ServerApp] nbdime | extension was successfully linked.
[I 2020-09-17 17:27:45.179 ServerApp] nbclassic | extension was successfully linked.
[W 2020-09-17 17:27:45.210 ServerApp] All authentication is disabled.  Anyone who can connect to this server will be able to run code.
[I 2020-09-17 17:27:45.211 ServerApp] The port 8888 is already in use, trying another port.
[I 2020-09-17 17:27:45.216 ServerApp] elyra | extension was successfully loaded.
[I 2020-09-17 17:27:45.218 ServerApp] jupyterlab_git | extension was successfully loaded.
[I 2020-09-17 17:27:45.222 ServerApp] nbclassic | extension was successfully loaded.
[I 2020-09-17 17:27:45.304 ServerApp] nbdime | extension was successfully loaded.
[I 2020-09-17 17:27:45.304 ServerApp] Serving notebooks from local directory: /Users/kbates/repos/oss/gateway-experiments/jupyter_server
[I 2020-09-17 17:27:45.304 ServerApp] Jupyter Server 1.0.0rc16 is running at:
[I 2020-09-17 17:27:45.304 ServerApp] http://localhost:8889/
[I 2020-09-17 17:27:45.304 ServerApp]  or http://127.0.0.1:8889/
[I 2020-09-17 17:27:45.304 ServerApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
```

**Comment by kevin-bates** [comment]
I do see these warnings when running the tests...
```
tests/test_gateway.py::test_gateway_cli_options
  /opt/anaconda3/envs/jupyter-server-dev/lib/python3.7/site-packages/traitlets/traitlets.py:2198: FutureWarning: Supporting extra quotes around Unicode is deprecated in traitlets 5.0. Use 'http://mock-gateway-server:8889' instead of "'http://mock-gateway-server:8889'" – or use CUnicode.
    FutureWarning)

tests/test_gateway.py::test_gateway_cli_options
  /opt/anaconda3/envs/jupyter-server-dev/lib/python3.7/site-packages/traitlets/traitlets.py:2198: FutureWarning: Supporting extra quotes around Unicode is deprecated in traitlets 5.0. Use 'alice' instead of "'alice'" – or use CUnicode.
    FutureWarning)
```
But these can be addressed by [changing/fixing their composition](https://github.com/jupyter/jupyter_server/blob/master/tests/test_gateway.py#L158-L159) to remove the extra string indicator.

So is the reason for making this change then to preserve compatibility with existing configurations (files or CLI scripts) where the initialization includes redundant quotes that previously weren't an issue?

**Comment by martinRenou** [comment]
The warning comes from here: https://github.com/ipython/traitlets/blob/bf8bcc3f6df248a80a6f8c181c48e12917b9cb5e/traitlets/traitlets.py#L2196-L2199

> So is the reason for making this change then to preserve compatibility with existing configurations (files or CLI scripts) where the initialization includes redundant quotes that previously weren't an issue?

Yes, also it will soon prevent setting the value to an empty string, which we use in Voila tests.

**Review by minrk — CHANGES_REQUESTED** [review]
I think this is the opposite of the desired result from the changes in traitlets 5, and I think it is the result of a faulty recommendation in the warning coming from traitlets (removed https://github.com/ipython/traitlets/pull/635). Sorry about that!

The C* traits should be used very rarely (even less often with traitlets 5), and only to accept configuration that is *not* of the desired type. In particular, the changes in traitlets 5 mean that `--string-key=5` will no longer raise an error with unicode traits and will correctly result in the string `"5"` (CUnicode casts 5 to str, allowing that to work without quotes in traitlets 4 as well, the exact workaround that is no longer necessary in traitlets 5).

Adopting CUnicode as is done here means that you will have config like `c.JupyterServerApp.notebook_dir = <anyting-not-a-string>` become valid instead of raising an error on mistakes. I don't think that's an improvement.

**Comment by minrk** [comment]
> also it will soon prevent setting the value to an empty string

Not soon, at least not for some years. There are no plans for the deprecated behavior to ever stop working (we don't want traitlets changes to break *anything*).

**Comment by minrk** [comment]
I should also clarify that it is not accurate that setting empty strings on the command-line is unsupported. Only `--config='""'`, where you are *not* passing an empty strong, you are passing the two-character string `'""'`, is deprecated as a way to specify an empty string. For actual empty strings, `--string-trait=` should work the same on traitlets 4 and 5, setting an empty string from an actual empty string.

The only place where the two behaviors should differ for strings is on: `--string-trait=5` or `--string-trait=[]` where a Pyton-literal interpretation of the value would result in a type that is not a string. In traitlets 4 this results in an error (hence the need for double-quoting) in traitlets 5 the text is preserved as intended removing the need for double-quoting, though that behavior, while deprecated, is preserved indefinitely for backward compatibility.

**Comment by Zsailer** [comment]
I'm going to close this, given @minrk's feedback in https://github.com/jupyter-server/jupyter_server/pull/306#pullrequestreview-509256977

If something further needs to be done, feel free to reopen.

**Pull request closed without merging**